PKCS11- Support for Key Generation (Generic Secret and AES Keys)

ruchi393 opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -711,9 +714,51 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 	/* If class/type not defined, match from mechanism */
 	if (get_class(temp) == PKCS11_UNDEFINED_ID &&
```
I did quick visual comparison to what we have in our code base for this PR and it seems that this needs some refactoring here about how "guessing" the intent of caller is resolved -- I can try to bake a PR for that.

However I do not see that the change will affect the outcome for this PR but will affect the ones coming later.

ruchi393 replied:
@vesajaaskelainen , I would agree that guessing part can be cleaned up here. We decided to leave as it is for now and work on it in another PR. If you have the changes available at your end, please feel free to raise a PR for same once this PR is merged.

vesajaaskelainen replied:
I am OK having that after this PR. We have tested our changes with quite a good set of tools and so far the updated logic seems to be OK. Let's finetune the feature in future PRs while we go forward. In meanwhile I try to rebase all of our changes to track upstream state to make it easier to collaborate.

ruchi393 replied:
Thanks @vesajaaskelainen . Can you please also help review this and other PR's associated with Key generation in optee_client and optee_test repos.

etienne-lms replied:
A this stage, these 2 mechanisms mandate the generated object class and type.
So I think here we should first guess the object type (secret key gen. =>CKK_GEN._SECRET, aes key gen.=>CKK_AES, otherwise, CKK_UNDEFINED) and change `sanitize_consistent_class_and_type()` to get 2 new input arguments: class and type, possibly `PKCS11_CKx_UNDEFINED`. The function current use this later value as default class and type when first finding these attributes.

		case PKCS11_CKM_GENERIC_SECRET_KEY_GEN:
		goto out;				class = PKCS11_CKO_SECRET_KEY;
			type = PKCS11_CKK_GENERIC_SECRET;
			break;
		case PKCS11_CKM_AES_KEY_GEN:
			class = PKCS11_CKO_SECRET_KEY;
			type = PKCS11_CKK_AES;

ruchi393 replied:
I agree with you @etienne-lms . Do you want me to make this change as part of this PR or a separate PR ?

etienne-lms replied:
i think you can add to this P-R.

vesajaaskelainen replied:
I have code snippets already for the guessing part so that same semantics work for different use cases too. Do you want that I take this PR and add commit on top of that so you can see how it looks?

etienne-lms replied:
You can create a P-R (even if it duplicates this one) or simply share a github commit url on your tree, as you like.

vesajaaskelainen replied:
I will try to port more commits to upstream tree, but here are the changes for this:
https://github.com/vesajaaskelainen/optee_os/commit/91f895166b500c3cdf1b105cc59eeb40eda603e8

I did this on top of @ruchi393's work and whole branch is here:
https://github.com/vesajaaskelainen/optee_os/commits/pkcs11-fix-mecha-attr-contribution

Example snippet for EC and RSA:
```
@@ -692,9 +692,11 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
                        type = PKCS11_CKK_AES;
                        break;
                case PKCS11_CKM_EC_KEY_PAIR_GEN:
+                       class = template_class;
                        type = PKCS11_CKK_DH;
                        break;
                case PKCS11_CKM_RSA_PKCS_KEY_PAIR_GEN:
+                       class = template_class;
                        type = PKCS11_CKK_RSA;
                        break;
                default:
```

```
@@ -421,7 +422,8 @@ enum pkcs11_rc entry_generate_key_pair(struct pkcs11_client *client,
        rc = create_attributes_from_template(&pub_head,
                                             template, template_size, NULL,
                                             PKCS11_FUNCTION_GENERATE_PAIR,
-                                            proc_params->id);
+                                            proc_params->id,
+                                            PKCS11_CKO_PUBLIC_KEY);
        if (rc)
                goto out;
 
@@ -442,7 +444,8 @@ enum pkcs11_rc entry_generate_key_pair(struct pkcs11_client *client,
        rc = create_attributes_from_template(&priv_head,
                                             template, template_size, NULL,
                                             PKCS11_FUNCTION_GENERATE_PAIR,
-                                            proc_params->id);
+                                            proc_params->id,
+                                            PKCS11_CKO_PRIVATE_KEY);
        if (rc)
                goto out;
 
@@ -844,7 +847,8 @@ enum pkcs11_rc entry_derive_key(struct pkcs11_client *client,
        rc = create_attributes_from_template(&head, template, template_size,
                                             parent_obj->attributes,
                                             PKCS11_FUNCTION_DERIVE,
-                                            proc_params->id);
+                                            proc_params->id,
+                                            PKCS11_CKO_UNDEFINED_ID);
        if (rc)
                goto out;
```

vesajaaskelainen replied:
I was able to make "ugly" version running in "upstream" tree that is now in github:
https://github.com/vesajaaskelainen/optee_os/tree/ugly-tree-starting-point

It is a bit mixed tree and does not follow upstream but is combination of @etienne-lms's tree and then EC improvement support patch by Ricardo Salveti and then lots of fixes while testing with different tools & suites and by reading the specs. There is still work ahead to pass test suites. So far the tools are happy (except the trust policy part of NSS's certutil -- which is rejected)

I am happy to assist in getting things in place but perhaps this PR is not the place. Feel free to propose method.

I also started writing up wiki page how one can test things:
https://github.com/vesajaaskelainen/optee_os/wiki/Testing-thigs

vesajaaskelainen replied:
@etienne-lms + @ruchi393 : shall we create a issue for integrating pkcs11 support in so that we can discuss there more in details -- eg. what I can take from my tree and so on?

ruchi393 replied:
Sure, that sounds good. We have our Monthly OP-TEE meeting planned on Monday (Dec 21). If you are planning to join that meeting, we can use that as starting point for our discussion.

At high level, I am currently working to upstream the code in the mainline optee-os by using sks branch from etienne's tree in the order given below :
1. C_GenerateKey (this PR)
2. I have started work on the remaining Object API's - C_FindObjects*, C_GetAttribute, C_CopyObject
3. Next  in plan is C_GenerateKeyPair (RSA)

It would be good to collaborate and work in parallel on this stuff to get things in mainstream sooner.

vesajaaskelainen replied:
I did not know that such meeting exists. I believe I can join for online meeting on next Monday. Please do send the details to my email.

ruchi393 replied:
You can find the details on:
https://www.linaro.org/projects/#LOC

Meeting detail is available at 
https://www.trustedfirmware.org/meetings/

ruchi393 commented on ta/pkcs11/src/token_capabilities.c:
```diff
@@ -238,6 +242,11 @@ void mechanism_supported_key_sizes(uint32_t proc_id, uint32_t *min_key_size,
 {
 	switch (proc_id) {
+	case PKCS11_CKM_GENERIC_SECRET_KEY_GEN:
+	/* This mechanism expects the keysize to be returned in bits */
```
Fixed

ruchi393 commented on ta/pkcs11/src/processing.c:
```diff
@@ -110,6 +110,178 @@ size_t get_object_key_bit_size(struct pkcs11_object *obj)
+	/*
+	 * Now obj_handle (through the related struct pkcs11_object instance)
+	 * owns the serialized buffer that holds the object attributes.
+	 * We reset attrs->buffer to NULL as serializer object is no more
```
Changed

ruchi393 commented on ta/pkcs11/src/attributes.c:
```diff
@@ -58,6 +58,49 @@ enum pkcs11_rc add_attribute(struct obj_attrs **head, uint32_t attribute,
+	cur = (char *)h + sizeof(struct obj_attrs);
+	end = cur + h->attrs_size;
+	for (; cur < end; cur += next_off) {
+		struct pkcs11_attribute_head pkcs11_ref;
```
Fixed

ruchi393 commented on ta/pkcs11/include/pkcs11_ta.h:
```diff
@@ -423,6 +423,21 @@ enum pkcs11_ta_cmd {
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [out] memref[2] = 32bit object handle
+	 *
+	 * This command relates to the PKCS#11 API functions C_GenerateKey().
```
Changed

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -784,8 +843,12 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 		/* Keys mandate attribute PKCS11_CKA_KEY_GEN_MECHANISM */
-		mechanism_id = PKCS11_CK_UNAVAILABLE_INFORMATION;
+			mechanism_id = PKCS11_CK_UNAVAILABLE_INFORMATION;
+		}
```
can remove `{`/`}`:
```c
		if (local)
			mechanism_id = mecha;
		else
			mechanism_id = PKCS11_CK_UNAVAILABLE_INFORMATION;

ruchi393 replied:
Fixed

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -938,11 +1001,29 @@ enum pkcs11_rc check_created_attrs_against_processing(uint32_t proc_id,
 
+	switch (proc_id) {
+	case PKCS11_CKM_GENERIC_SECRET_KEY_GEN:
+		if (get_key_type(head) != PKCS11_CKK_GENERIC_SECRET)
```
could be asserted since implementation checks/mandates these.

ruchi393 replied:
Agreed. We are already checking for this in create_attributes_from_template() and returning error from there. So, here we can just add an assert to check of this mismatch is still there. I will make the change.

ruchi393 commented:
I have taken care of the review comments except the one for sanitize_consistent_class_and_type().
@vesajaaskelainen , let me know how you want to proceed with  that.

Apart from that I have added another commit in the PR -  "ta: pkcs11: Fix for CKA_LOCAL in create_attributes_from_template()" has been added to fix CKA_LOCAL attribute check issue which I found during testing.

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -790,7 +790,7 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 		goto out;
 
-	if (get_attribute(attrs, PKCS11_CKA_LOCAL, NULL, NULL) !=
+	if (get_attribute(temp, PKCS11_CKA_LOCAL, NULL, NULL) !=
```
I looks it should be the same for `PKCS11_CKA_KEY_GEN_MECHANISM` at line 799, no? think

ruchi393 replied:
Yes, you are right. As I was reading the spec, similar restriction would also apply to attributes like CKA_ALWAYS_SENSITIVE and CKA_NEVER_EXTRACTABLE. Basically for all attributes marked with restrictions 2,4,6 in Table 10.  I will add the checks and corresponding testcases for these.

etienne-lms commented on ta/pkcs11/src/processing.c:
```diff
@@ -110,6 +110,178 @@ size_t get_object_key_bit_size(struct pkcs11_object *obj)
+						PKCS11_FUNCTION_GENERATE,
+						PKCS11_FUNC_STEP_INIT);
+	if (rc) {
+		DMSG("Mechanism not supported %x %x", proc_params->id, rc);
```
`PRIx32` for `proc_params->id`.
(rephrased to fit in 80chars)
```c
		DMSG("Invalid mechanism %#"PRIx32": %#x", proc_params->id, rc);

vesajaaskelainen commented:
> I have taken care of the review comments except the one for sanitize_consistent_class_and_type().
> @vesajaaskelainen , let me know how you want to proceed with that.
> 
> Apart from that I have added another commit in the PR - "ta: pkcs11: Fix for CKA_LOCAL in create_attributes_from_template()" has been added to fix CKA_LOCAL attribute check issue which I found during testing.

Please see the commit I referred in more in place..

ruchi393 commented:
@etienne-lms , @vesajaaskelainen , please check the last patch marked with RFC. I think we can solve the issue of template without class and type by passing this as a hint to sanitize_object_class(). This hint if passed can be added to the list of attributes if class and type are not already present in the original template.

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -707,13 +707,13 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
+	 * already present
+	 */
+	if (function == PKCS11_FUNCTION_GENERATE) {
 		switch (mecha) {
```
Can't add comment above in function signature ;)

But please remove '__unused.

ruchi393 replied:
Sure, I will remove that.  I have been planning to do so whenever I see that but somehow it gets skipped.

ruchi393 replied:
Done in latest PR

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -724,24 +724,18 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 			type = PKCS11_CKK_AES;
 			break;
 		default:
```
What we are missing here is of course asymmetric keys support.

In that we have PKCS11_CKO_PUBLIC_KEY or PKCS11_CKO_PRIVATE_KEY depending which template is being processed.

For this reason I also added template_class to method signature so that detail can be carried to path where it gets added to actual object.

vesajaaskelainen replied:
One could probably implement it in here if missing then set class = template_class; and then carry on to sanitize_client_object.

ruchi393 replied:
I agree with you. 

As we add GenerateKeyPair in the picture, we can add the template_class as you have suggested in create_attributes_from_template() and further based on the mecha, class can be template class and type would come from mecha.

case PKCS11_CKM_EC_KEY_PAIR_GEN:
			class = template_class;
			type = PKCS11_CKK_EC;
			break;
case PKCS11_CKM_RSA_PKCS_KEY_PAIR_GEN:
			class = template_class;
			type = PKCS11_CKK_RSA;
			break;

ruchi393 replied:
Added class_id as unused param in latest PR.

ruchi393 commented:
@etienne-lms , @vesajaaskelainen , I have taken care of the review comments till date. Please review.

For the remaining changes not affecting GenerateKey() but being used in GenerateKeyPair(), I would suggest we take care of them in PR adding support for GenerateKeyPair().

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -745,19 +785,22 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 		goto out;
 
-	if (get_attribute(attrs, PKCS11_CKA_LOCAL, NULL, NULL) !=
+	if (get_attribute(temp, PKCS11_CKA_LOCAL, NULL, NULL) !=
```
Just a note -- no need to change if you feel current is better:

If you are not interested in attribute's value at all then `get_attribute_ptr(temp, PKCS11_CKA_LOCAL, NULL, NULL)` would be slightly more lightweight but `get_attribute()` also gets the job done.

ruchi393 replied:
Good idea. I will change this.

vesajaaskelainen commented:
> @etienne-lms , @vesajaaskelainen , I have taken care of the review comments till date. Please review.
> 
> For the remaining changes not affecting GenerateKey() but being used in GenerateKeyPair(), I would suggest we take care of them in PR adding support for GenerateKeyPair().

No need to tackle RSA and EC key stuff in this PR.

ruchi393 commented:
@etienne-lms , @vesajaaskelainen , should I squash the commits and update the PR.

vesajaaskelainen commented:
> @etienne-lms , @vesajaaskelainen , should I squash the commits and update the PR.

I am OK with that.

ruchi393 commented:
Rebased and squashed the patches for easier review. All comments till date have been taken care of.

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -684,7 +684,8 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 				struct obj_attrs *parent __unused,
 				enum processing_func function,
-				enum pkcs11_mechanism_id mecha __unused)
+				enum pkcs11_mechanism_id mecha,
```
Removal of the __unsed from mecha should be in commit https://github.com/OP-TEE/optee_os/pull/4257/commits/a1ef9669dae5b1ef1ac89f9f5b5ea4fb9c349705 as that starts to use the argument.

ruchi393 replied:
Done

vesajaaskelainen commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -724,24 +724,18 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 			type = PKCS11_CKK_AES;
 			break;
 		default:
-			EMSG("Unable to define class/type from mechanism");
```
Now we lost trap for unsupported mechanism in here?

ruchi393 replied:
Thanks for catching this. I will add a TEE_Panic over here. The unsupported mechanism check is already done in optee_client. So if we have reached this default condition the something has gone wrong.

vesajaaskelainen replied:
I checked out PR https://github.com/OP-TEE/optee_client/pull/241 and didn't notice any checks for that.

I believe original code is quite good and gives nice error to caller:
```c
		default:
			EMSG("Unable to define class/type from mechanism");
			rc = PKCS11_CKR_TEMPLATE_INCOMPLETE;
			goto out;
```

In my opinion this is good error handling in this case and doesn't cause any havoc anywhere.

We could probably give different error thou... perhaps `PKCS11_CKR_MECHANISM_INVALID`?

ruchi393 replied:
serialize_ck_mecha_params() in libckteec/src/serialize_ck.c does the check in optee_client code.

vesajaaskelainen replied:
Ok. 

I believe the pre-check in `serialize_ck_mecha_params()` is OK.

However -- I don't think you can so easily do `TEE_Panic()`.

`TEE_Panic()` should only be done when there is grave error in within TA code.

Also it is not strictly required that someone uses libckteec for communication. Eg. trust layer should be entry to the TA.

In here we are validating input from REE and we can easily bail out. I do not see this being in grave error category.

vesajaaskelainen commented:
@ruchi393 Went thru the commits.

For commits:
https://github.com/OP-TEE/optee_os/pull/4257/commits/67747b085c1b99d9e95fdf92efe229cc03949fd2
https://github.com/OP-TEE/optee_os/pull/4257/commits/80ae8736d5a6ea58fac70b0a4cd8c306dec43d82
https://github.com/OP-TEE/optee_os/pull/4257/commits/e29df73cb2ca66441291d68f5cc85343053b41fd
https://github.com/OP-TEE/optee_os/pull/4257/commits/fa72a1960ced75a026297b3cfc10ccda57bd68de

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

Tiny cleanup comment:
https://github.com/OP-TEE/optee_os/pull/4257/commits/08d29ca406cf5bc82d70620ec2c3c74267b39f14

This looks good but referred by the change in previous comment:
https://github.com/OP-TEE/optee_os/pull/4257/commits/a1ef9669dae5b1ef1ac89f9f5b5ea4fb9c349705

This needs a bit cleanup as noticed by check-patch:
https://github.com/OP-TEE/optee_os/pull/4257/commits/57855e9772fa6743634b3c3f2a8593c503cc82b6

Added comment to:
https://github.com/OP-TEE/optee_os/pull/4257/commits/f4eace0f44bdf39018f7af34a3fef83bb29484f6

ruchi393 commented:
Thanks for the review @vesajaaskelainen . I have done the following:
- added tags in commits you mentioned
-  fixed checkpatch error and __unused commnet
-  added additional patch for review to take care of lost error for mechanism.

vesajaaskelainen commented:
For commits:

https://github.com/OP-TEE/optee_os/pull/4257/commits/a5c35ed3532bb34195d47d6121bfa9c714229905
https://github.com/OP-TEE/optee_os/pull/4257/commits/c9a888c7cee92bc3f72d574af9e62b97b708b916

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

If you squash it to original commit it is OK or you can leave it as separate:

https://github.com/OP-TEE/optee_os/pull/4257/commits/93957fd9aae525dd27c324abbf02159a7cede748

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

This https://github.com/OP-TEE/optee_os/pull/4257/commits/99168dba9c85130b230141230d21bbd0e7962050 should be squashed into
https://github.com/OP-TEE/optee_os/pull/4257/commits/64b9bafe221c6af047f12de0e55e8b7e9d366e60. And then for that:

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

ruchi393 commented:
@vesajaaskelainen , thanks a lot for helping in review of the patches.

Can you also please review the PR in optee_client -
https://github.com/OP-TEE/optee_client/pull/241

ruchi393 commented:
Squashed and tags applied. Thanks @vesajaaskelainen
